src/libutil/config.hh

```diff
@@ -214,17 +214,49 @@ protected:
     bool isOverridden() const;
 };
 
+/**
+ * Settings, like command-line arguments, have a "handler" which is
+ * executed when the value is set, including the default value.
+ * The handler may cause arbitrary side-effects.
+ */
+template<typename T>
+struct SettingHandler
+{
+    /**
+     * The `bool` argument is `true` if the value is being set to the default
+     * value.
+     *
+     * The `T &` argument is the value itself.
+     */
+    std::function<void(bool, T &)> fun;
+
+    SettingHandler() = default;
+
+    SettingHandler(std::function<void(bool, T &)> && fun)
+        : fun(std::move(fun))
+    { }
+
+    operator bool() const {
+        if (fun) {
+            return true;
+        } else {
+            return false;
+        }
```

Comment:
```suggestion
        return fun;
```

?

Reply:
Weirdly the boolean coercion doesn't work here. I tried this but it wouldn't compile.

Reply:
```
src/libutil/config.hh:240:16: error: no viable conversion from returned value of type 'const std::function<void (bool, basic_string<char, char_traits<char>, allocator<char>> &)>' to function return type 'bool'
        return fun;
               ^~~
src/libutil/config.hh:313:13: note: in instantiation of member function 'nix::SettingHandler<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>::operator bool' requested here
        if (handler) {
            ^
src/libutil/config.hh:386:45: note: in instantiation of member function 'nix::Setting<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>::assign' requested here
    void operator =(const Path & v) { this->assign(v); }
                                            ^
/nix/store/pkbcb46dxzpkf4zhkl094cn08ql7jh4h-libcxx-11.1.0-dev/include/c++/v1/functional:2406:22: note: explicit conversion function is not a candidate
    _LIBCPP_EXPLICIT operator bool() const _NOEXCEPT {
```

Reply:
Okay, two notes:

1. `static_cast<bool>(fun)` did seem to work.
2. If we set a default value for `fun`, we don't need this conversion operator at all.